sys/fido2/ctap/ctap_cbor.c

```diff
@@ -0,0 +1,1691 @@
+            return ret;
+        }
+
+        ret = cbor_value_advance(&map);
+        if (ret != CborNoError) {
+            return CTAP2_ERR_CBOR_PARSING;
+        }
+    }
+
+    if (required_parsed != 5) {
+        return CTAP2_ERR_MISSING_PARAMETER;
+    }
+
+    return CTAP2_OK;
+}
+
+/* parse PublicKeyCredentialRpEntity dictionary */
+static int parse_rp(CborValue *it, ctap_rp_ent_t *rp)
+{
+    int ret;
+    int type;
+    CborValue map;
+    size_t map_len;
+    size_t key_len;
+    size_t len;
+    char key[8];
+    uint8_t required_parsed = 0;
+
+    type = cbor_value_get_type(it);
+    if (type != CborMapType) {
+        return CTAP2_ERR_INVALID_CBOR_TYPE;
+    }
+
+    ret = cbor_value_enter_container(it, &map);
+    if (ret != CborNoError) {
+        return CTAP2_ERR_CBOR_PARSING;
+    }
+
+    ret = cbor_value_get_map_length(it, &map_len);
+    if (ret != CborNoError) {
+        return CTAP2_ERR_CBOR_PARSING;
+    }
+
+    for (size_t i = 0; i < map_len; i++) {
+        type = cbor_value_get_type(&map);
+        if (type != CborTextStringType) {
+            return CTAP2_ERR_INVALID_CBOR_TYPE;
+        }
+
+        key_len = sizeof(key);
+        ret = cbor_value_copy_text_string(&map, key, &key_len, NULL);
+        if (ret == CborErrorOutOfMemory) {
+            return CTAP2_ERR_LIMIT_EXCEEDED;
+        }
+
+        key[sizeof(key) - 1] = 0;
+
+        ret = cbor_value_advance(&map);
+        if (ret != CborNoError) {
+            return CTAP2_ERR_CBOR_PARSING;
+        }
+
+        type = cbor_value_get_type(&map);
+        if (type != CborTextStringType) {
+            return CTAP2_ERR_INVALID_CBOR_TYPE;
+        }
+
+        if (strcmp(key, "id") == 0) {
+            rp->id_len = CTAP_DOMAIN_NAME_MAX_SIZE;
+            ret =
+                parse_text_string(&map, (char *)rp->id, (size_t *)&rp->id_len);
+            if (ret != CborNoError) {
+                return ret;
+            }
+
+            required_parsed++;
+        }
+        else if (strcmp(key, "name") == 0) {
+            len = CTAP_RP_MAX_NAME_SIZE;
+            ret = parse_text_string(&map, (char *)rp->name, &len);
+            if (ret != CborNoError) {
+                return ret;
+            }
+        }
+        else if (strcmp(key, "icon") == 0) {
+            len = CTAP_DOMAIN_NAME_MAX_SIZE;
+            ret = parse_text_string(&map, (char *)rp->icon, &len);
+            if (ret != 0) {
+                return ret;
+            }
+        }
+        else {
+            DEBUG("CTAP_parse_rp: ignoring unknown key: %s \n", key);
+        }
+
+        ret = cbor_value_advance(&map);
+        if (ret != CborNoError) {
+            return CTAP2_ERR_CBOR_PARSING;
+        }
+    }
+
+    /* id is mandatory */
+    if (required_parsed != 1) {
+        return CTAP2_ERR_MISSING_PARAMETER;
+    }
+
+    return CTAP2_OK;
+}
+
+/* parse PublicKeyCredentialUserEntity dictionary */
+static int parse_user(CborValue *it, ctap_user_ent_t *user)
```

Comment:
I could put common parsing functionality of `parse_rp` and `parse_user` into a function but this would make the code less readable and additional logic would be needed to differ between if `ctap_user_ent_t` or `ctap_rp_ent_t` is being parsed.

Reply:
Please try. The are very similar so it is worth it. The if/else paths are implemented a little differently but I think many cases can be harmonized (e.g.,  `if (ret != CborNoError)` vs `if (ret != CTAP2_OK)`, `else if (strncmp(key, CTAP_CBOR_DISPLAY_NAME, strlen(CTAP_CBOR_DISPLAY_NAME)) == 0)` vs `else if (strcmp(key, CTAP_CBOR_STR_NAME) == 0)` ...)

Reply:
I combined the methods. Is it better now ?